File: 01/node/tyranny.js
const fs = require('fs');
const path = require('path');  
const readline = require('readline');  
const inputPath = path.join(__dirname, 'input');


function part1(total, moduleMass){
    return total += fuelNeeded(moduleMass)
}

function part2(total, moduleMass){
    return fuelNeeded(moduleMass) > 0 ? part2(part1(total, moduleMass), fuelNeeded(moduleMass)) : total
}


async function computeFuelNeeded(compute = part1){
    const rl = readline.createInterface({
        input: fs.createReadStream(inputPath),
    });
    let total = 0;
    for await (const line of rl) {
        total = compute(total, line)
    }
    return total
}

function fuelNeeded(mass){
    return Math.floor(((Math.abs(mass/3)))) - 2
}


console.log(computeFuelNeeded(part1).then(t => console.log(t)))
console.log(computeFuelNeeded(part2).then(t => console.log(t)))

File: 01/python/tyranny.py
def part1():
    masses = open("input")
    total = 0
    for mass in masses:
        total += fuel(mass)
    return total

def part2():
    masses = open("input")
    total = 0
    for mass in masses:
        reqfuel = fuel(mass)
        total += reqfuel
        reqfuel = fuel(reqfuel)
        while (reqfuel > 0):
            total += reqfuel
            reqfuel = fuel(reqfuel)
    return total

def fuel(mass):
    return int(abs(int(mass)/3.0))-2

print part1()
print part2()
